{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Problem3.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pachterlab/Bi-BE-CS-183-2022/blob/main/HW1/Problem3.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Bi/Be/Cs 183 2021-2022: Intro to Computational Biology\n",
        "TAs: Meichen Fang, Tara Chari, Zitong (Jerry) Wang\n",
        "\n",
        "**Submit your notebooks by sharing a clickable link with Viewer access. Link must be accessible from submitted assignment document.**\n",
        "\n",
        "Make sure Runtime $\\rightarrow$ Restart and run all works without error"
      ],
      "metadata": {
        "id": "JPUswnzT0EZG"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**HW 1 Problem 3**\n",
        "\n",
        "For this problem you will be exploring the phenomenon known as Simpson's Reversal in a single-cell sequencing dataset.\n",
        " \n",
        "As described in Lecture, Simpson's Reversal describes the phenomenon where a certain trend appears within certain groups of the data, but disappears or is reversed when the groupings are changed. One of the advantages of single-cell RNA sequencing as opposed to bulk RNA sequencing (where all reads from all cells are pooled together) is that it provides the opportunity, in principle, to avoid incorrect inferences about correlation between pairs of genes.\n",
        "\n",
        "For this problem you will mine a single-cell RNA-seq dataset to see if this phenomenon occurs in practice."
      ],
      "metadata": {
        "id": "VthQpDn4tt96"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_tty4AYi77ys"
      },
      "source": [
        "#To run a code cell, select the cell and hit Command/Ctrl+Enter or click the run/play symbol\n",
        "#Click Insert --> Code Cell or the '+ Code' option to insert a new code cell"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BAjFr_0y8Sxb"
      },
      "source": [
        "#Click Insert --> Text Cell or the '+ Text' option to insert a cell for text as below"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# This is  used to time the running of the notebook\n",
        "import time\n",
        "start_time = time.time()"
      ],
      "metadata": {
        "id": "Nj7fi7kx05Mu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MozW3XxY8ezr"
      },
      "source": [
        "Text here for descriptions, explanations, etc"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5oPRpu328neb"
      },
      "source": [
        "##**Import data and install packages**"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "vYTZEjs9AqwV"
      },
      "source": [
        "import numpy as np\n",
        "import scipy.io as sio\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt #Can use other plotting packages like seaborn"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kNFuUFJ78cN5"
      },
      "source": [
        "# ! allows you to run commands in the command line, as you would in your normal terminal/command line interface"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yH-EpSmB9Y_x",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e600eec6-5719-4f22-8d93-590d67e1f81c"
      },
      "source": [
        "#Download count matrix of cell by gene counts\n",
        "#Cells are 10x sequenced neurons from the mouse hypothalamus (Kim et al. 2019)\n",
        "\n",
        "!wget --content-disposition https://data.caltech.edu/tindfiles/serve/c5db82c2-4052-439e-a3f1-38acd89f469e/"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-13 17:34:15--  https://data.caltech.edu/tindfiles/serve/c5db82c2-4052-439e-a3f1-38acd89f469e/\n",
            "Resolving data.caltech.edu (data.caltech.edu)... 34.250.186.131, 34.252.192.177, 34.251.243.51\n",
            "Connecting to data.caltech.edu (data.caltech.edu)|34.250.186.131|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 FOUND\n",
            "Location: https://tind-caltechdata.s3.amazonaws.com:443/d05a9752c6084408b1d91671edb446a2?Signature=BRa43EVSHtYGEhiQEuwSWd252nM%3D&Expires=1639417156&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-type=application/force-download&response-content-disposition=attachment%3Bfilename%3D%22tenx_sub.mtx.gz%22 [following]\n",
            "--2021-12-13 17:34:16--  https://tind-caltechdata.s3.amazonaws.com/d05a9752c6084408b1d91671edb446a2?Signature=BRa43EVSHtYGEhiQEuwSWd252nM%3D&Expires=1639417156&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-type=application/force-download&response-content-disposition=attachment%3Bfilename%3D%22tenx_sub.mtx.gz%22\n",
            "Resolving tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)... 52.218.57.98\n",
            "Connecting to tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)|52.218.57.98|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 30671193 (29M) [application/force-download]\n",
            "Saving to: ‘tenx_sub.mtx.gz’\n",
            "\n",
            "tenx_sub.mtx.gz     100%[===================>]  29.25M  15.5MB/s    in 1.9s    \n",
            "\n",
            "2021-12-13 17:34:18 (15.5 MB/s) - ‘tenx_sub.mtx.gz’ saved [30671193/30671193]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lBDp8sjR_fsQ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c092ee6f-4add-4e86-d617-c0915d30c99a"
      },
      "source": [
        "#Download csv of metadata for cells (cell type classification, sex of mouse) \n",
        "\n",
        "!wget --content-disposition https://data.caltech.edu/tindfiles/serve/11e5d726-f921-4bc9-9c13-d0d8712c185d/"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-13 17:34:19--  https://data.caltech.edu/tindfiles/serve/11e5d726-f921-4bc9-9c13-d0d8712c185d/\n",
            "Resolving data.caltech.edu (data.caltech.edu)... 34.250.186.131, 34.252.192.177, 34.251.243.51\n",
            "Connecting to data.caltech.edu (data.caltech.edu)|34.250.186.131|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 FOUND\n",
            "Location: https://tind-caltechdata.s3.amazonaws.com:443/3a24d928727e4355ac0d8f58d1ec1fd6?Signature=g8pnLMMrFF%2FLibHsQ%2F32cvz%2Brrs%3D&Expires=1639417159&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-type=application/force-download&response-content-disposition=attachment%3Bfilename%3D%22tenx_obs_sub.csv.gz%22 [following]\n",
            "--2021-12-13 17:34:19--  https://tind-caltechdata.s3.amazonaws.com/3a24d928727e4355ac0d8f58d1ec1fd6?Signature=g8pnLMMrFF%2FLibHsQ%2F32cvz%2Brrs%3D&Expires=1639417159&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-type=application/force-download&response-content-disposition=attachment%3Bfilename%3D%22tenx_obs_sub.csv.gz%22\n",
            "Resolving tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)... 52.218.57.98\n",
            "Connecting to tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)|52.218.57.98|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 75077 (73K) [application/force-download]\n",
            "Saving to: ‘tenx_obs_sub.csv.gz’\n",
            "\n",
            "tenx_obs_sub.csv.gz 100%[===================>]  73.32K   353KB/s    in 0.2s    \n",
            "\n",
            "2021-12-13 17:34:20 (353 KB/s) - ‘tenx_obs_sub.csv.gz’ saved [75077/75077]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "khusZj08PuxC",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "59a9e0d3-2318-40cd-b605-d011f3d22335"
      },
      "source": [
        "#Download csv of metadata for genes (gene names)\n",
        "\n",
        "!wget --content-disposition https://data.caltech.edu/tindfiles/serve/0582f31f-b700-4a9b-8f6d-790a79abc74c/"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-13 17:34:20--  https://data.caltech.edu/tindfiles/serve/0582f31f-b700-4a9b-8f6d-790a79abc74c/\n",
            "Resolving data.caltech.edu (data.caltech.edu)... 34.250.186.131, 34.252.192.177, 34.251.243.51\n",
            "Connecting to data.caltech.edu (data.caltech.edu)|34.250.186.131|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 FOUND\n",
            "Location: https://tind-caltechdata.s3.amazonaws.com:443/8eb77ef038e3427582851162df61e760?Signature=xmBi%2FAnSEbe3kEXd53tVT%2FOMNxU%3D&Expires=1639417160&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-type=application/force-download&response-content-disposition=attachment%3Bfilename%3D%22var%20%281%29.csv.gz%22 [following]\n",
            "--2021-12-13 17:34:20--  https://tind-caltechdata.s3.amazonaws.com/8eb77ef038e3427582851162df61e760?Signature=xmBi%2FAnSEbe3kEXd53tVT%2FOMNxU%3D&Expires=1639417160&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-type=application/force-download&response-content-disposition=attachment%3Bfilename%3D%22var%20%281%29.csv.gz%22\n",
            "Resolving tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)... 52.218.57.98\n",
            "Connecting to tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)|52.218.57.98|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 13141 (13K) [application/force-download]\n",
            "Saving to: ‘var (1).csv.gz’\n",
            "\n",
            "var (1).csv.gz      100%[===================>]  12.83K  --.-KB/s    in 0s      \n",
            "\n",
            "2021-12-13 17:34:21 (75.5 MB/s) - ‘var (1).csv.gz’ saved [13141/13141]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xrGNmEHwAKAF"
      },
      "source": [
        "#Gunzip the files we downloaded\n",
        "!gunzip *.gz"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PPv4FIlm8rtQ"
      },
      "source": [
        "## **Read in data for analysis**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bldcPDtroHZ_"
      },
      "source": [
        "**The dataset**\n",
        "\n",
        "This dataset is taken from [Kim et al. 2019](https://www.sciencedirect.com/science/article/pii/S0092867419310712). The goal of this study was to discern the various neuronal cell types present in the mouse ventromedial hypothalamus (VMH), a region whose neurons control/influence social behaviors e.g. mounting and aggression. \n",
        "\n",
        "This dataset comprises neurons from *male and female* mice in different behavioral states, covering *30 distinct neuron populations or cell types*.\n",
        "\n",
        "Here we are analyzing only the 10x sequenced samples.\n",
        "\n",
        "<img src=\"https://ars.els-cdn.com/content/image/1-s2.0-S0092867419310712-fx1_lrg.jpg\" alt=\"AbstractFigure\" width=\"500\" height=\"500\">\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2qq9TO5V97Vr"
      },
      "source": [
        "**The Count matrix**\n",
        "\n",
        "This matrix is 11,591 cells by 1,999 genes. The full dataset contains 41,580 cells, however we will work with a subsetted version to facilitate calculations within the Colab environment.\n",
        "\n",
        "1.   For each cell, gene counts were normalized to have the same number of total counts (usually 1e5 or 1e6), with cell-gene counts thus scaled accordingly.\n",
        "\n",
        "2.   Counts were then log-normalized, using the log(1+x), where x is each cell's gene count. The 1 accounts for 0 count genes. \n",
        "\n",
        "3. The ~2000 genes were selected for those that displayed large variance in expression amongst the cells ('highly variable genes').\n",
        "\n",
        "\n",
        " \n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tZ1h_dME8vJb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "406e6ac0-e5f7-43be-81a2-5e279e2da274"
      },
      "source": [
        "#Get gene count matrix\n",
        "count_mat = sio.mmread('tenx_sub.mtx')\n",
        "count_mat.shape"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(11591, 1999)"
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "NFissu8VB92P",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "outputId": "b3aa964f-f647-44cd-9d1a-71da3da32171"
      },
      "source": [
        "#Get metadata dataframe for the 11,591 cells (rows of the matrix)\n",
        "meta = pd.read_csv('tenx_obs_sub.csv',index_col = 0)\n",
        "meta.head()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>cell_type</th>\n",
              "      <th>sex_label</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>cell_barcode</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>4_AAACCTGCACAGTCGC</th>\n",
              "      <td>Tsix_Esr1_1</td>\n",
              "      <td>F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4_AAACCTGCATACTCTT</th>\n",
              "      <td>Tsix_Esr1_1</td>\n",
              "      <td>F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4_AAACCTGGTCCGTTAA</th>\n",
              "      <td>Scgn</td>\n",
              "      <td>F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4_AAACCTGGTTACGGAG</th>\n",
              "      <td>Esr1_1</td>\n",
              "      <td>F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4_AAACCTGTCAATCACG</th>\n",
              "      <td>Tsix_Esr1_1</td>\n",
              "      <td>F</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                      cell_type sex_label\n",
              "cell_barcode                             \n",
              "4_AAACCTGCACAGTCGC  Tsix_Esr1_1         F\n",
              "4_AAACCTGCATACTCTT  Tsix_Esr1_1         F\n",
              "4_AAACCTGGTCCGTTAA         Scgn         F\n",
              "4_AAACCTGGTTACGGAG       Esr1_1         F\n",
              "4_AAACCTGTCAATCACG  Tsix_Esr1_1         F"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "M5je0CyVN4DN",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "outputId": "a3f907b8-a9bc-495d-c343-daef542bdc67"
      },
      "source": [
        "#Get metadata dataframe for the 1,999 genes (columns of the matrix)\n",
        "\n",
        "meta_gene = pd.read_csv('var (1).csv',index_col = 0)\n",
        "meta_gene.head()\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>Gm1992</th>\n",
              "      <td>Gm1992</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Lypla1</th>\n",
              "      <td>Lypla1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Npbwr1</th>\n",
              "      <td>Npbwr1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Sntg1</th>\n",
              "      <td>Sntg1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1700034P13Rik</th>\n",
              "      <td>1700034P13Rik</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                   gene_name\n",
              "Gm1992                Gm1992\n",
              "Lypla1                Lypla1\n",
              "Npbwr1                Npbwr1\n",
              "Sntg1                  Sntg1\n",
              "1700034P13Rik  1700034P13Rik"
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xzz-5cVyCsS4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "085b9fa6-531c-45dc-d8a1-a5f4db03e50d"
      },
      "source": [
        "#Access the list in each column by meta['name_of_col']\n",
        "cell_types = meta['cell_type']\n",
        "cell_types[0:5]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "cell_barcode\n",
              "4_AAACCTGCACAGTCGC    Tsix_Esr1_1\n",
              "4_AAACCTGCATACTCTT    Tsix_Esr1_1\n",
              "4_AAACCTGGTCCGTTAA           Scgn\n",
              "4_AAACCTGGTTACGGAG         Esr1_1\n",
              "4_AAACCTGTCAATCACG    Tsix_Esr1_1\n",
              "Name: cell_type, dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "plqTILR-DdnC",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f889bd36-2aaf-40a3-95f4-0e50dc4bfc72"
      },
      "source": [
        "#See all the cell type categories\n",
        "np.unique(meta['cell_type']) #Cell populations are marked by the genes they express (as comparec to other cell populations)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Dlk1_1', 'Esr1_1', 'Nr5a1_1', 'Nr5a1_Foxp2_1', 'Nup62cl',\n",
              "       'Satb2_1', 'Scgn', 'Tsix_Esr1_1'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-kcBQFk_p3bC",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7018f9bc-65f2-4902-a08e-73047475d814"
      },
      "source": [
        "#Filter the count matrix for a particular cell type (across all genes)\n",
        "\n",
        "#Remember! Cells are ROWS and Genes are COLUMNS\n",
        "\n",
        "indices = meta['cell_type'].isin(['Tsix_Esr1_1']) #Get indices for where the cell_type is Dlk1_1 only\n",
        "\n",
        "count_sub = count_mat[indices,:] #Filter rows (cells) of matrix\n",
        "count_sub.shape"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(1659, 1999)"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xRCYW1VVErG1",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f8a64ddb-277c-4366-8a08-c89f99dbb052"
      },
      "source": [
        "#The sexes\n",
        "np.unique(meta['sex_label'])"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['F', 'M'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gPKnb4VNEJE9"
      },
      "source": [
        "## **Problem 3 (36 points)** \n",
        "Find an example of Simpson's reversal."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AfyMXQRkE32t"
      },
      "source": [
        "Generally Simpson's reversal describes how a trend may appear in several groups of data, but not exist or be reversed when groups are combined.\n",
        "\n",
        "Here we will focus on trends of correlation, where two variables (genes) appear correlated in one direction (e.g. **positively**) given one set of groupings, but display **no or negative** correlation in another set of groupings."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "z28ojhm7HIEl"
      },
      "source": [
        "For the purposes of this homework, we will be using the common Pearson correlation coefficient to measure correlation between variables.\n",
        "\n",
        "This can be calculated using numpy's corrcoef() function in python, where correlation values are between -1 and 1."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "NkHBhRNOEP5X",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "92e662d7-5c36-4e9d-d817-e1fccb044fc6"
      },
      "source": [
        "#Use np.corrcoef()\n",
        "\n",
        "#Given two lists/vectors of values x and y, np.corrcoef(x,y) will return a \n",
        "#pairwise correlation matrix where the diagonal represent correlation between\n",
        "#x with x or y with y, (which are 1) and the off-diagonals which both represent  \n",
        "#correlation of x with y. \n",
        "\n",
        "#The off-diagonal value is the value we are interested in\n",
        "\n",
        "x = [1.0,2.0,3.0,4.0]\n",
        "y = [4.0,5.0,2.0,6.0]\n",
        "\n",
        "corrs = np.corrcoef(x,y)\n",
        "print('Correlation matrix: \\n', corrs)\n",
        "\n",
        "print('Corr in first row, second column: ', corrs[0,1])"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Correlation matrix: \n",
            " [[1.         0.22677868]\n",
            " [0.22677868 1.        ]]\n",
            "Corr in first row, second column:  0.22677868380553634\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WSk8kc65MhXR"
      },
      "source": [
        "**Problem:** Find two genes which display a Simpson's reversal in their correlations depending on how they are grouped. We are calculating correlation between the genes' expression vectors i.e. their values across cells.\n",
        "\n",
        "Examples of grouping comparison include, but are not limited to:\n",
        "\n",
        "\n",
        "*   Correlation of two genes within one cell type versus their correlation across all cells (no cell type grouping)\n",
        "*   Correlation of two genes within one cell type versus their correlations within each sex within the same cell type\n",
        "*   Correlation of two genes in each sex versus across all cells (no sex grouping)\n",
        "\n",
        "\n",
        "*Try to report the most extreme difference in correlation you can find*"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "03eGGOJeNRc9"
      },
      "source": [
        "Show:\n",
        "\n",
        "a.   All code for subsetting the count matrix and calculating correlation values for a pair of genes. (12 points)\n",
        "\n",
        "Code can be added to code cells below. Feel free to add more code cells if necessary.\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "K4crXALfvCUu",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "150ad571-21fd-4071-8ede-bd1f72a1ba76"
      },
      "source": [
        "# ------- Example code for finding gene-pair correlations within a cell type versus across all cells -----------\n",
        "\n",
        "gene1 = 'Lypla1' #Lysophospholipase (a hydrolase)\n",
        "gene2 = 'H2-K1'  #Histocompatibility antigen, Involved in presentation of antigens to immune system\n",
        "\n",
        "indices = meta['cell_type'].isin(['Tsix_Esr1_1']) #Get indices for where the cell_type is Dlk1_2 only\n",
        "\n",
        "g1_indices = meta_gene['gene_name'].tolist().index(gene1) #Get exact index for where gene names are gene1 only\n",
        "g2_indices = meta_gene['gene_name'].tolist().index(gene2)\n",
        "\n",
        "#Correlation of counts within cell type\n",
        "count_sub_g1 = count_mat[indices,g1_indices] #Filter rows (cells) of matrix\n",
        "count_sub_g2 = count_mat[indices,g2_indices]\n",
        "\n",
        "corrs = np.corrcoef(count_sub_g1,count_sub_g2)\n",
        "print('Correlation within type: ', corrs[0,1])\n",
        "\n",
        "#Correlation of counts within cell type\n",
        "count_sub_g1 = count_mat[:,g1_indices] #Filter rows (cells) of matrix Sntg1\n",
        "count_sub_g2 = count_mat[:,g2_indices]\n",
        "\n",
        "corrs = np.corrcoef(count_sub_g1,count_sub_g2)\n",
        "print('Correlation across all cells: ',corrs[0,1])\n",
        "\n",
        "#We see slight positive correlation within the Tsix_Esr1 cell type but almost no correlation across all cells"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Correlation within type:  0.35349248871444433\n",
            "Correlation across all cells:  0.029026343554515218\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RcvmDytf4gX0"
      },
      "source": [
        "# ------- Your code below ------- \n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "p0qGbExH4gcg"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4tIii_pT4gjS"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "iKnihYR0kqCH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "SX70L58akqE2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "BQwrLBqHaQYp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "AaydHS8fYlq1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "bgnxu2laaqhq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zXIabplmuw-U"
      },
      "source": [
        "Report: \n",
        "\n",
        "b.   What the comparison groupings are (e.g. within one cell type versus across all cells) (6 points)\n",
        "\n",
        "c.   Which two genes you are looking at (6 points)\n",
        "\n",
        "d.   What the gene-gene correlations are in each grouping (6 points)\n",
        "\n",
        "e.   Qualitatively comment on how significant or relevant you think these correlation differences are, given the groupings you chose and the correlation values. (6 points)\n",
        "\n",
        "*Answers b-e can be added directly to this text box*\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Running time of the notebook\n",
        "print(\"{:.2f} minutes\".format((time.time()-start_time)/60))"
      ],
      "metadata": {
        "id": "U-KtIXUr09w1",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "91df756e-4411-4a4f-c17c-1d9a45fec834"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "0.61 minutes\n"
          ]
        }
      ]
    }
  ]
}